server/commands/connected_objects.go

```diff
@@ -170,65 +175,74 @@ func (c *ConnectedObjectsCommand) reverseExpandTupleToUserset(
 		},
 	})
 	if err != nil {
+		iter1.Stop()
 		return err
 	}
 
 	iter := storage.NewCombinedIterator(iter1, iter2)
+	defer iter.Stop()
 
 	g := errgroup.Group{}
 	g.SetLimit(100) // set some concurrency limit
 
+ForLoop:
 	for {
-		t, err := iter.Next()
-		if err != nil {
-			if errors.Is(err, storage.ErrIteratorDone) {
-				break
+		select {
+		case <-ctx.Done():
+			// if the request times out, or it's cancelled manually, we short circuit so that we avoid unnecessary database lookups
+			break ForLoop
+		default:
+			t, err := iter.Next()
+			if err != nil {
+				if errors.Is(err, storage.ErrIteratorDone) {
+					break ForLoop
+				}
+
+				return err
 			}
 
-			return err
-		}
+			tk := t.GetKey()
 
-		tk := t.GetKey()
+			foundObject := tk.GetObject()
+			foundObjectType, _ := tuple.SplitObject(foundObject)
 
-		foundObject := tk.GetObject()
-		foundObjectType, _ := tuple.SplitObject(foundObject)
+			userObj, _ := tuple.SplitObjectRelation(tk.GetUser())
 
-		userObj, _ := tuple.SplitObjectRelation(tk.GetUser())
+			if userObj == tuple.Wildcard {
 
-		if userObj == tuple.Wildcard {
+				return serverErrors.InvalidTuple(
+					fmt.Sprintf("unexpected wildcard evaluated on relation '%s#%s'", foundObjectType, tuplesetRelation),
+					tuple.NewTupleKey(foundObject, tuplesetRelation, tuple.Wildcard),
+				)
+			}
 
-			return serverErrors.InvalidTuple(
-				fmt.Sprintf("unexpected wildcard evaluated on relation '%s#%s'", foundObjectType, tuplesetRelation),
-				tuple.NewTupleKey(foundObject, tuplesetRelation, tuple.Wildcard),
-			)
-		}
+			if _, ok := foundObjectsMap.Load(foundObject); ok {
+				// todo(jon-whit): we could optimize this by avoiding reading this
+				// from the database in the first place
 
-		if _, ok := foundObjectsMap.Load(foundObject); ok {
-			// todo(jon-whit): we could optimize this by avoiding reading this
-			// from the database in the first place
+				// if we've already evaluated/found the object, then continue
+				continue
+			}
 
-			// if we've already evaluated/found the object, then continue
-			continue
-		}
+			if foundObjectType == sourceObjectType {
+				if foundCount != nil && atomic.AddUint32(foundCount, 1) > c.Limit {
+					break ForLoop
+				}
 
-		if foundObjectType == sourceObjectType {
-			if foundCount != nil && atomic.AddUint32(foundCount, 1) > c.Limit {
-				break
+				resultChan <- foundObject
+				foundObjectsMap.Store(foundObject, struct{}{})
 			}
 
-			resultChan <- foundObject
-			foundObjectsMap.Store(foundObject, struct{}{})
+			g.Go(func() error {
```

Comment:
Following along from my comment in https://github.com/openfga/openfga/pull/368/files#r1039832386

Here we would derive a sub group off of the top-level errgroup. For example,

```
var parentCtx context.Context // comes from request or caller
...

topgroup, gctx := errgroup.WithContext(parentCtx)

for _, ingress := range ingresses {
    switch ingress.Type {
    case graph.DirectIngress:
        err = c.reverseExpandDirect(gctx, topgroup, ...)
    ...
}

func (c *ConnectedObjectsCommand) reverseExpandDirect(ctx context.Context, g *errgroup.Group, ...) {
    ...
    subg, subgctx := errgroup.WithContext(ctx)
    
    for {
        ...
        subg.Go(func() error {...})
    }

    return subg.Wait()
}
```